app/src/lib/stores/app-store.ts

```diff
@@ -3417,19 +3426,150 @@ export class AppStore extends TypedBaseStore<IAppState> {
     return this._refreshRepository(repository)
   }
 
+  /** This shouldn't be called directly. See `Dispatcher`. */
+  public _setRebaseProgressFromState = async (repository: Repository) => {
+    const snapshot = await getRebaseSnapshot(repository)
+    if (snapshot === null) {
+      return
+    }
+
+    const { progress, commits } = snapshot
+
+    this.repositoryStateCache.updateRebaseState(repository, () => {
+      return {
+        progress,
+        commits,
+      }
+    })
+  }
+
+  /** This shouldn't be called directly. See `Dispatcher`. */
+  public _initializeRebaseProgress(
+    repository: Repository,
+    commits: ReadonlyArray<CommitOneLine>
+  ) {
+    this.repositoryStateCache.updateRebaseState(repository, () => {
+      const hasCommits = commits.length > 0
+      const firstCommitSummary = hasCommits ? commits[0].summary : null
+
+      return {
+        progress: {
+          value: formatRebaseValue(0),
+          rebasedCommitCount: 0,
+          currentCommitSummary: firstCommitSummary,
+          totalCommitCount: commits.length,
+        },
+        commits,
+      }
+    })
+
+    this.emitUpdate()
+  }
+
+  /** This shouldn't be called directly. See `Dispatcher`. */
+  public _setConflictsResolved(repository: Repository) {
+    // an update is not emitted here because there is no need
+    // to trigger a re-render at this point
+
+    this.repositoryStateCache.updateRebaseState(repository, () => ({
+      userHasResolvedConflicts: true,
+    }))
+  }
+
+  /** This shouldn't be called directly. See `Dispatcher`. */
+  public async _setRebaseFlowStep(
+    repository: Repository,
+    step: RebaseFlowStep
+  ): Promise<void> {
+    this.repositoryStateCache.updateRebaseState(repository, () => ({
+      step,
+    }))
+
+    this.emitUpdate()
+
+    if (step.kind === RebaseStep.ShowProgress && step.rebaseAction !== null) {
+      await timeout(500)
```

Comment:
Why the timeout of 500ms? If this isn't an arbitrary value, it may be worthing throwing it in a constant.

Reply:
This was an unscientific number chosen to allow the UI time to display itself before the next action was performed (which could be very fast). Digging into some more data on this:

> The average reaction time for humans is 0.25 seconds to a visual stimulus, 0.17 for an audio stimulus, and 0.15 seconds for a touch stimulus.
>
> [[source](https://backyardbrains.com/experiments/reactiontime)]

So it's a generous delay that gives users a better chance to see the rebase progress before it switches back to the conflicts UI (if it encounters an issue).

I'm not sure how much this value might change, or get reused in other places, but I'll add a comment in here to explain this area a bit better.

Reply:
I've added more context behind this to e8aace4b90f54ca640af4806811ab40cbfc955af